doc/design/service-discovery.md

```diff
@@ -0,0 +1,316 @@
+potential issues
+* Fast due to shared memory
+
+Con:
+
+* Shared memory approach might have overhead compared to a simple RPC call
+
+#### Event-based notification
+
+##### Alternative A: Shared memory + IPC channel
+
+* Send ID's over IPC channel and read ABA-counter/ServiceDiscoveryChangeCounter from shared memory
+* Data transfer over shared memory
+  * Basically putting Service Registry to shared memory so everyone can access it (read-only)
+
+Pro:
+
+* Fast data transfer
+* Could replace `InterfacePort`s
+
+Contra:
+
+* Bespoken solution more complex than todays approach (thread-safety required, lock-free mechanism needed)
+* In an optimal case IPC channel communication shall only be used for startup
+
+##### Alternative B: Built-in topic based on `InterfacePort`'s
+
+* Built-in topic approach based on `InterfacePort`'s
+  * Sniff and intercept `CaproMessage`
+* Also see
+  * [overview diagram](diagrams/service-discovery-sequence-diagram.puml)
+  * [sequence diagram](diagrams/service-discovery-overview.puml)
+
+Pro:
+
+* Close to DDS mechanism
+* `InterfacePort` already available
+* Synergies with gateway mechanisms
+* Gateways can benefit from event notification mechanism
+  * No need for a polling-loop anymore
+
+Con:
+
+* Overhead, `Listner` will wake up on every `CaproMessage`
+* Bookkeeping at user-side, just delta of service registry is transferred
+* What will RouDi do if he runs out of memory?
+  * Dimensioning according to max values is not optimal (MAX_INTERFACE_CAPRO_FIFO_SIZE)
+  * Presumably lots of memory needed, e.g. during startup phase when lots of apps will do discovery
+  * A safety-certified middleware based on iceoryx would not use the dynamic discovery feature
+
+Remark:
+
+* Need to filter out your own `CaproMessageType::FIND` requests
+* Extend `dispatchCaproMessage()` with condition variable and notification mechanism
+
+##### Alternative C: Custom thread
+
+* Implement a custom thread in `PoshRuntime`, which polls for specific discovery information using request/response
+
+Pro:
+
+* More flexible, e.g. thread can be used for other functionality in future
+* Re-use req/resp building blocks
+
+Con:
+
+* Not needed by all users per default
+* Polling leads to overhead
+
+##### Alternative D: Discovery Class + Listner
+
+Create a new publisher in RouDi which sends a `ServiceRegistryTopic`. This publisher would be used both to signal a
+change in the service registry and to transmit the service discovery registry. The complete old service registry
+(saved locally) would be compared to the new service registry in a new class, extending the public user API.
+
+Pro:
+
+* Simple and consistent user API for both event-based and sychronous requests
+  * Filtering for `findService` could be done inside the new class
+* No need for a change counter as ABA problem solved by POSH mechanisms
+* Not just delta of service registry is transmitted, but complete service registry info
```

Comment:
From my point of view this can even be improved later to also base the current cregistry state on incremental updates in addition. In some way it is a state estimation problem with the known registry being the state and some information (full state or updates) arriving which can be used to reestimate the state.